File: src/python/day1/puzzle2.py
from collections import defaultdict

def similarityScore(a, b):
    counts = defaultdict(int)
    for n in b:
        counts[n]+=1
    return sum([n*counts[n] for n in a if counts[n]])

if __name__ == "__main__":
    print(similarityScore([1, 2], [2, 1]))

File: src/go/day1/puzzle2.go
package main

import "fmt"

func puzzle2() {
	fmt.Println(similarityScore(a1, a2))
}

func similarityScore(a1, a2 []int) int {
	a2Counts := make(map[int]int, len(a1))
	for i := 0; i < len(a1); i++ {
		a2Counts[a2[i]]++
	}
	sum := 0
	for _, n := range a1 {
		sum += n * a2Counts[n]
	}
	return sum
}
